File: python/semantic_kernel/core_skills/conversation_summary_skill.py
# Copyright (c) Microsoft. All rights reserved.
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from semantic_kernel.kernel import Kernel
    from semantic_kernel.orchestration.sk_context import SKContext


class ConversationSummarySkill:
    """
    Semantic skill that enables conversations summarization.
    """

    from semantic_kernel.skill_definition import sk_function

    # The max tokens to process in a single semantic function call.
    _max_tokens = 1024

    _summarize_conversation_prompt_template = (
        "BEGIN CONTENT TO SUMMARIZE:\n{{"
        + "$INPUT"
        + "}}\nEND CONTENT TO SUMMARIZE.\nSummarize the conversation in 'CONTENT TO"
        " SUMMARIZE',            identifying main points of discussion and any"
        " conclusions that were reached.\nDo not incorporate other general"
        " knowledge.\nSummary is in plain text, in complete sentences, with no markup"
        " or tags.\n\nBEGIN SUMMARY:\n"
    )

    def __init__(self, kernel: "Kernel"):
        self._summarizeConversationFunction = kernel.create_semantic_function(
            ConversationSummarySkill._summarize_conversation_prompt_template,
            skill_name=ConversationSummarySkill.__name__,
            description=(
                "Given a section of a conversation transcript, summarize the part of"
                " the conversation."
            ),
            max_tokens=ConversationSummarySkill._max_tokens,
            temperature=0.1,
            top_p=0.5,
        )

    @sk_function(
        description="Given a long conversation transcript, summarize the conversation.",
        name="SummarizeConversation",
        input_description="A long conversation transcript.",
    )
    async def summarize_conversation_async(
        self, input: str, context: "SKContext"
    ) -> "SKContext":
        """
        Given a long conversation transcript, summarize the conversation.

        :param input: A long conversation transcript.
        :param context: The SKContext for function execution.
        :return: SKContext with the summarized conversation result.
        """
        from semantic_kernel.text import text_chunker
        from semantic_kernel.text.function_extension import (
            aggregate_chunked_results_async,
        )

        lines = text_chunker._split_text_lines(
            input, ConversationSummarySkill._max_tokens, True
        )
        paragraphs = text_chunker._split_text_paragraph(
            lines, ConversationSummarySkill._max_tokens
        )

        return await aggregate_chunked_results_async(
            self._summarizeConversationFunction, paragraphs, context
        )

File: java/semantickernel-core-skills/src/main/java/com/microsoft/semantickernel/coreskills/ConversationSummarySkill.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.coreskills;

import com.microsoft.semantickernel.Kernel;
import com.microsoft.semantickernel.builders.SKBuilders;
import com.microsoft.semantickernel.orchestration.SKContext;
import com.microsoft.semantickernel.skilldefinition.annotations.DefineSKFunction;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionInputAttribute;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionParameters;
import com.microsoft.semantickernel.text.TextChunker;
import com.microsoft.semantickernel.textcompletion.CompletionSKFunction;

import reactor.core.publisher.Mono;

import java.util.List;

import javax.annotation.Nullable;

/** Semantic skill that enables conversations summarization */
public class ConversationSummarySkill {
    // The max tokens to process in a single semantic function call.
    private static final int MaxTokens = 1024;

    private final CompletionSKFunction summarizeConversationFunction;
    private final CompletionSKFunction conversationActionItemsFunction;
    private final CompletionSKFunction conversationTopicsFunction;
    private final Kernel kernel;

    /**
     * Initializes a new instance of the ConversationSummarySkill class
     *
     * @param kernel Kernel instance
     */
    public ConversationSummarySkill(Kernel kernel) {
        this.kernel = kernel;
        this.summarizeConversationFunction =
                SKBuilders.completionFunctions(kernel)
                        .createFunction(
                                SemanticFunctionConstants.SummarizeConversationDefinition,
                                "summarize",
                                "ConversationSummarySkill",
                                "Given a section of a conversation transcript, summarize the part"
                                        + " of the conversation.",
                                SKBuilders.completionConfig()
                                        .maxTokens(MaxTokens)
                                        .temperature(0.1)
                                        .topP(0.5)
                                        .build());

        this.conversationActionItemsFunction =
                SKBuilders.completionFunctions(kernel)
                        .createFunction(
                                SemanticFunctionConstants.GetConversationActionItemsDefinition,
                                "ActionItems",
                                "ConversationSummarySkill",
                                "Given a section of a conversation transcript, identify action"
                                        + " items.",
                                SKBuilders.completionConfig()
                                        .maxTokens(MaxTokens)
                                        .temperature(0.1)
                                        .topP(0.5)
                                        .build());

        this.conversationTopicsFunction =
                SKBuilders.completionFunctions(kernel)
                        .createFunction(
                                SemanticFunctionConstants.GetConversationTopicsDefinition,
                                "Topics",
                                "ConversationSummarySkill",
                                "Analyze a conversation transcript and extract key topics worth"
                                        + " remembering.",
                                SKBuilders.completionConfig()
                                        .maxTokens(MaxTokens)
                                        .temperature(0.1)
                                        .topP(0.5)
                                        .build());
    }

    /**
     * Given a long conversation transcript, summarize the conversation
     *
     * @param input A long conversation transcript
     * @param context The SKContext for function execution
     * @return A context containing the summarized conversation
     */
    @DefineSKFunction(
            description = "Given a long conversation transcript, summarize the conversation.",
            name = "SummarizeConversation")
    public Mono<SKContext> summarizeConversationAsync(
            @SKFunctionInputAttribute
                    @SKFunctionParameters(
                            description = "A long conversation transcript.",
                            name = "input")
                    String input,
            @Nullable SKContext context) {
        List<String> lines = TextChunker.splitPlainTextLines(input, MaxTokens);
        List<String> paragraphs = TextChunker.splitPlainTextParagraphs(lines, MaxTokens);

        if (context == null) {
            context =
                    SKBuilders.context().with(kernel.getSkills()).with(kernel.getMemory()).build();
        }

        SKContext completionContext = context.copy();

        return this.summarizeConversationFunction.aggregatePartitionedResultsAsync(
                paragraphs, completionContext);
    }

    /**
     * Given a long conversation transcript, identify action items
     *
     * @param input A long conversation transcript
     * @param context The SKContext for function execution
     * @return A context containing the action items
     */
    @DefineSKFunction(
            description = "Given a long conversation transcript, identify action items.",
            name = "GetConversationActionItems")
    public Mono<SKContext> getConversationActionItemsAsync(
            @SKFunctionInputAttribute
                    @SKFunctionParameters(
                            description = "A long conversation transcript.",
                            name = "input")
                    String input,
            SKContext context) {
        List<String> lines = TextChunker.splitPlainTextLines(input, MaxTokens);
        List<String> paragraphs = TextChunker.splitPlainTextParagraphs(lines, MaxTokens);

        SKContext completionContext = context.copy();

        return this.conversationActionItemsFunction.aggregatePartitionedResultsAsync(
                paragraphs, completionContext);
    }

    /**
     * Given a long conversation transcript, identify topics
     *
     * @param input A long conversation transcript
     * @param context The SKContext for function execution
     * @return A context containing the topics
     */
    @DefineSKFunction(
            description =
                    "Given a long conversation transcript, identify topics worth remembering.",
            name = "GetConversationTopics")
    public Mono<SKContext> getConversationTopicsAsync(
            @SKFunctionInputAttribute
                    @SKFunctionParameters(
                            description = "A long conversation transcript.",
                            name = "input")
                    String input,
            SKContext context) {
        List<String> lines = TextChunker.splitPlainTextLines(input, MaxTokens);
        List<String> paragraphs = TextChunker.splitPlainTextParagraphs(lines, MaxTokens);

        SKContext completionContext = context.copy();

        return this.conversationTopicsFunction.aggregatePartitionedResultsAsync(
                paragraphs, completionContext);
    }
}
